### DISCUSSION TITLE
Guidance sought: Creating a Lambda Layer via CDK using C#/.NET

### DISCUSSION THREAD

**Discussion by jnpwly** [aws/aws-cdk#23414] [General]
Kia ora, AWS CDK! 🇳🇿 

I am wanting to use the CDK to describe a new Lambda Layer, which can then be used by a number of subsequent Lambdas within our stack. We are a .NET shop, so we instinctively want to do this work with .NET. However, I am finding it difficult to find examples of:

* coding a Lambda Layer in C#, and then
* defining the Lambda Layer in a CDK Project using C#, and then
* deploying this stack via `cdk deploy`

Do such examples exist?

The closest I can find include:

* this: https://github.com/aws/aws-extensions-for-dotnet-cli/blob/master/docs/Layers.md#layers-for-net-core-lambda-functions, 
* this: https://github.com/awsimaya/Lambda-Layers, and
* this: https://aws.amazon.com/blogs/developer/aws-lambda-layers-with-net-core/

but they all use the `Amazon.Lambda.Tools` .NET Global Tool to deploy a Lambda Layer direct to an S3 Bucket, and do not use the CDK.

I am able to code up a Lambda in C# and then get that Lambda to be referenced/defined in a C# CDK Project, similar to this:

```csharp
            Function function = new Function(this,
               "Email-Lambda-Generate-HooblyHooEmail",
               new FunctionProps
               {
                   Runtime = Runtime.DOTNET_6,
                   Code = Code.FromAsset("../src/Email.Lambda/Email.Lambda.Generate.HooblyHooEmail"),
                   Handler = "HooblyHooEmail.Lambda::HooblyHooEmail.Lambda.Function::FunctionHandler"
               });
```

When executing `cdk deploy`, the source directory gets zipped up, uploaded to an S3 Bucket, and the Lambda is created. Great.

It seems, however, than the automatic creation of a Lambda Layer -- specifically, with the generation of NuGet packages -- does not happen via `cdk deploy`. This then means that manual steps are required to get the source into the desired state before executing `cdk deploy`. See: https://stackoverflow.com/questions/72992860/how-to-correctly-deploy-a-dotnet-lambda-layer-with-aws-cdk#comment128928520_72992860

If anyone can point me in the direction of an example that shows this has been done, that would be awesome.

Many thanks,
   Jon

**Comment by vishal-shah-25** [comment]
Hello,

I'm currently encountering the same issue. I am eager to establish a Lambda layer in C# within .NET 6.0, wherein I have incorporated various shared methods for AWS services like S3, SNS, etc. The process entails the inclusion of NuGet packages for S3 and SNS, such as AWSSDK.S3.dll.

By executing the CLI command **"dotnet lambda publish-layer <layer-name> --layer-type runtime-package-store --s3-bucket bucket-name"** I can generate the requisite NuGet packages. This command provides a Layer ARN, which can be used during the deployment of the Lambda function.

layer-type **runtime-package-store**
The runtime package store layer is typically located under the /opt/runtime-package-store directory by default.

To enable the .NET Core runtime in Lambda to locate the NuGet assemblies from the layer, it is imperative to configure the **DOTNET_SHARED_STORE** environment variable to point to the paths where the runtime package store layer will be extracted.

For more detailed information, please refer to this [link](https://github.com/aws/aws-extensions-for-dotnet-cli/blob/master/docs/Layers.md).

While using AWS CDK, I have yet to come across a method to deploy, create, or download NuGet packages for the Lambda layer in a specified location, akin to the procedure explained using the CLI command.

1. Layer deployed via CLI and Lambda function is deployed via CDK providing environment variable "DOTNET_SHARED_STORE" then - **_Working fine_**
2. Layer deployed via CDK and Lambda function deployed via CDK providing environment variable "DOTNET_SHARED_STORE" then - **_Not working_**

Sample code for reference as below:

    const TestS3Layer = new lambda.LayerVersion(this, 'TestS3Layer', {
      code: lambda.Code.fromAsset('G:/POC/TestS3Layer/packages.zip'), // Can refer code here but not working...
      compatibleRuntimes: [lambda.Runtime.DOTNET_6],
      description: 'Layer created via AWs CDK.',
    });

    // Existing IAM role ARN
    const existingRoleArn = 'arn:aws:iam::1234567890:role/service-role/NET6_Runtime_POC-role';
    
    // Create an instance of the existing IAM role
    const existingRole = iam.Role.fromRoleArn(this, 'NET6_Runtime_POC-role', existingRoleArn);

    // Create lambda functio
    const myFunction = new lambda.Function(this, 'TestLambdaLayerIntegrationFunction', {
      runtime: lambda.Runtime.DOTNET_6,
      handler: 'TestLambdaLayerIntegrationFunction::TestLambdaLayerIntegrationFunction.Function::FunctionHandler',
      code: lambda.Code.fromAsset('G:/POC/TestLambdaLayerIntegrationFunction/src/TestLambdaLayerIntegrationFunction/bin/Release/net6.0/'),
      memorySize: 256,
      timeout: cdk.Duration.seconds(60),
      role: existingRole,
      environment: {
        DOTNET_SHARED_STORE: '/opt/dotnetcore/store/',
      },
      description: ".Net 6 Lambda function deployed using AWS CDK.",
      layers: [TestS3Layer]
    });


If anyone could kindly provide a reference link or a code example detailing how to achieve these tasks within AWS CDK, it would be greatly appreciated!

Thank you,
Vishal

@aws-cdk-automation
